controllers/storage.go

```diff
@@ -145,6 +146,123 @@ func (s *Storage) RemoveAllButCurrent(artifact sourcev1.Artifact) ([]string, err
+		expired := diff > ttl
+		if !info.IsDir() && info.Mode()&os.ModeSymlink != os.ModeSymlink {
+			if path != localPath && expired {
+				garbageFiles = append(garbageFiles, path)
+			}
+			totalFiles += 1
+			filesWithCreatedTs[info.ModTime().UTC()] = path
+		}
+		return nil
+	})
+	if len(errors) > 0 {
+		return nil, fmt.Errorf("can't walk over file: %s", strings.Join(errors, ","))
+	}
+
+	// We already collected enough garbage files to satisfy the no of max
+	// items that are supposed to be retained, so exit early.
+	if totalFiles-len(garbageFiles) < maxItemsToBeRetained {
+		return garbageFiles, nil
+	}
+
+	creationTimestamps := []time.Time{}
+	for ts := range filesWithCreatedTs {
+		creationTimestamps = append(creationTimestamps, ts)
+	}
+	// sort all timestamps in an ascending order.
+	sort.Slice(creationTimestamps, func(i, j int) bool { return creationTimestamps[i].Before(creationTimestamps[j]) })
+	for _, ts := range creationTimestamps {
+		path, ok := filesWithCreatedTs[ts]
+		if !ok {
+			return garbageFiles, fmt.Errorf("failed to fetch file for created ts: %v", ts)
+		}
+		sortedPaths = append(sortedPaths, path)
+	}
+
+	var collected int
+	noOfGarbageFiles := len(garbageFiles)
+	for _, path := range sortedPaths {
+		if path != localPath && !stringInSlice(path, garbageFiles) {
+			// If we previously collected a few garbage files with an expired ttl, then take that into account
+			// when checking whether we need to remove more files to satisfy the max no of items allowed
+			// in the filesystem, along with the no of files already removed in this loop.
+			if noOfGarbageFiles > 0 {
+				if (len(sortedPaths) - collected - len(garbageFiles)) > maxItemsToBeRetained {
+					// append path to garbageFiles and remove it from sortedPaths
+					garbageFiles = append(garbageFiles, path)
+					collected += 1
+				}
+			} else {
+				if len(sortedPaths)-collected > maxItemsToBeRetained {
+					garbageFiles = append(garbageFiles, path)
+					collected += 1
+				}
+			}
+		}
+	}
+
+	return garbageFiles, nil
+}
+
+// RemoveGarbageFiles removes all garabge files in the artifact dir according to the provided
+// retention options.
+func (s *Storage) RemoveGarbageFiles(artifact sourcev1.Artifact, maxItemsToBeRetained int, ttl time.Duration) ([]string, error) {
+	garbageFiles, err := s.getGarbageFiles(artifact, maxItemsToBeRetained, ttl)
+	if err != nil {
+		return nil, err
+	}
+	var errors []string
+	var deleted []string
+	if len(garbageFiles) > 0 {
+		for _, file := range garbageFiles {
+			err := os.Remove(file)
+			if err != nil {
+				errors = append(errors, err.Error())
+			} else {
+				deleted = append(deleted, file)
+			}
+		}
+	}
+	if len(errors) > 0 {
+		return deleted, fmt.Errorf("failed to remove files: %s", strings.Join(errors, " "))
+	}
+	return deleted, nil
+}
+
+func stringInSlice(a string, list []string) bool {
+	for _, b := range list {
+		if b == a {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
`sort.SearchStrings` may be handy to check whether a string is within a slice

Reply:
According to the docs:
> SearchStrings searches for x in a sorted slice of strings and returns the index
as specified by Search. The slice must be sorted in ascending order.

Don't think we can `sort.SearchStrings ` here